src/utils/account.ts

```diff
@@ -35,123 +40,183 @@ export const getAccount = async (args: {
+
+      const account = await unconnectedSmartWallet.connect({
+        client: thirdwebClient,
+        personalAccount: signerAccount,
+      });
+
+      return { account, adminAccount: signerAccount };
+    }
+
+    case WalletType.smartGcpKms: {
+      const signerAccount = await getGcpKmsAccount({
+        client: thirdwebClient,
+        name: walletDetails.gcpKmsResourcePath,
+        clientOptions: {
+          credentials: {
+            client_email: walletDetails.gcpApplicationCredentialEmail,
+            private_key: walletDetails.gcpApplicationCredentialPrivateKey,
+          },
+        },
+      });
+
+      const unconnectedSmartWallet = smartWallet({
+        chain,
+        sponsorGas: true,
+        factoryAddress: walletDetails.accountFactoryAddress ?? undefined,
+      });
+
+      const account = await unconnectedSmartWallet.connect({
+        client: thirdwebClient,
+        personalAccount: signerAccount,
+      });
+
+      return { account, adminAccount: signerAccount };
+    }
+
+    case WalletType.smartLocal: {
+      const adminAccount = await encryptedJsonToAccount(
+        walletDetails.encryptedJson,
+      );
+      const unconnectedSmartWallet = smartWallet({
+        chain,
+        sponsorGas: true,
+        factoryAddress: walletDetails.accountFactoryAddress ?? undefined,
+      });
+
+      const account = await unconnectedSmartWallet.connect({
+        client: thirdwebClient,
+        personalAccount: adminAccount,
+      });
+
+      return { account, adminAccount };
     }
     default:
       throw new Error(`Wallet type not supported: ${walletDetails.type}`);
   }
 };
 
+export const _adminAccountsCache = new Map<string, Account>();
+
+/**
+ * Get the admin account for a smart backend wallet (cached)
+ * Will throw if the wallet is not a smart backend wallet
+ */
+export const getSmartBackendWalletAdminAccount = async ({
+  chainId,
+  accountAddress,
+}: {
+  chainId: number;
+  accountAddress: Address;
+}) => {
+  const chain = await getChain(chainId);
+
+  // Get from cache.
+  const cacheKey = getAdminAccountCacheKey({ chainId, accountAddress });
+  const cached = _adminAccountsCache.get(cacheKey);
+  if (cached) {
+    return cached;
+  }
+
+  const walletDetails = await getWalletDetails({
+    address: accountAddress,
+  });
+
+  if (!isSmartBackendWallet(walletDetails)) {
+    throw new Error(
+      "Wallet is not a smart backend wallet and does not have an admin account",
+    );
+  }
+
+  const { adminAccount } = await walletDetailsToAccount({
+    walletDetails,
+    chain,
+  });
+
+  if (!adminAccount) {
+    // todo: error improvement, make it easy to parse whether user error or system error
+    throw new Error("Unexpected error: admin account not found");
+  }
+
+  _adminAccountsCache.set(cacheKey, adminAccount);
+  return adminAccount;
+};
+
+const getAdminAccountCacheKey = (args: {
+  chainId: number;
+  accountAddress: Address;
+}) => `${args.chainId}-${args.accountAddress}`;
```

Comment:
Shouldn't this be chain-agnostic if we store WalletDetails without chain? Though it doesn't hurt to include chain here too.